File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        /*
        
        [ 3, 6, 7, 11] h = 8
        k -> [1, 2, ..., 11]
        
        the max eating speed you could have is max(piles) per hour. 
        this is because if you went any higher, it wouldn't matter. 
        you still have to stay at each pile until the hour is finished.
        
        to choose the eating speed (k), you just have to try k values
        until you find one that takes the same amount of hours as h.
        
        as k increases, the total hours to finish decreases.
        */
        
        // return bruteForce(piles, h);
        return binSearch(piles, h);
    }
    
    public int binSearch(int[] piles, int h) {
        List<Integer> nums = new ArrayList<>();
        for (int n : piles) nums.add(n);
        int max = Collections.max(nums);

        int lo = 1; 
        int hi = max;
        while (lo < hi) {
            int speed = lo + ((hi - lo) / 2); // left mid.
            int totalHours = hoursToEat(piles, speed);

            if (h >= totalHours) {
                hi = speed;
            } else {
                lo = speed + 1;
            }
        }
        
        return lo;
    }
    
    public int bruteForce(int[] piles, int h) {
        List<Integer> nums = new ArrayList<>();
        for (int n : piles) nums.add(n);
        int max = Collections.max(nums);
        
        for (int k = 1; k <= max; ++k) {
            if (hoursToEat(piles, k) == h) return k;
        }
        
        return 0;
    }
    
    public int hoursToEat(int[] piles, double k) {
        int totalHours = 0;
        for (double count : piles) {
            totalHours += (int) Math.ceil(count / k);
        }
        
        return totalHours;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        """
        we are looking for k - we have to choose a k value.
        koko will sit there and take up the whole hour if the pile has < k bananas before moving to next pile.
        looking for the optimal speed/k so that you still eat ALL bananas in H hours.
        want the minimum k.
        
        ACCEPTABLE VALUES - k vals/speeds that let you finish less than h hours. (fast)
        UNACCEPTABLE VALUES - k vals/speeds that have you finish in more than h hours. (too slow)

        what is the range of k vals to search through?
        min: 1 banana/hr
        max: max(piles)/hr
        """

        def calculate_hours(speed: int):
            hours = 0
            for pile in piles:
                hours +=  ceil(pile / speed)
            return hours 

        lo = 1
        hi = max(piles)
        while lo < hi:
            speed = lo + ((hi - lo) // 2) # left mid
            hours_taken = calculate_hours(speed)

            if hours_taken <= h:
                hi = speed
            else:
                lo = speed + 1

        return lo


